numpy/_core/src/multiarray/refcount.c

```diff
@@ -56,6 +56,51 @@ PyArray_ClearBuffer(
 }
 
 
+/*
+ * Helper function to zero an array buffer.
+ *
+ * Here "zeroing" means an abstract zeroing operation,
+ * which for an array of references might be something
+ * more complicated than zero-filling the buffer.
+ *
+ * Failure (returns -1) indicates some sort of programming or
+ * logical error and should not happen for a data type that has
+ * been set up correctly.
+ */
+NPY_NO_EXPORT int
+PyArray_ZeroBuffer(
+        PyArray_Descr *descr, char *data,
+        npy_intp stride, npy_intp size, int aligned)
+{
+    if (!PyDataType_REFCHK(descr)) {
+        return 0;
```

Comment:
There are two problems here IMO:
* The refcheck isn't correct quite correct.  While I would say it is best if `memset(..., 0, ...)` is the bit pattern for the zeroed buffer, I don't think we need to encode that here (nor do we in the array creation path).
* With the refcheck the functino is in theory not fuzzy, but still not great IMO.  It should guarantee zeroing, but it doesn't.
  Doing that probably means making it `PyArray_ZeroContiguousBuffer()` so that `memset()` is trivial.  I suspect that is OK, this is mainly a helper after all.
  (And to proof that point, this part is currently incorrect in the code here.)

It is unfortunate that we can't re-use such a helper in the creation code since we need to malloc/calloc there, so the logic is reversed.  I don't see a good solution for that unless we want to put two helpers next to each other here.  Maybe just a comment there, that it mirrors this function?